Java profiler: Clarify allocation engine pre-requisits (#27640)
DataDog/documentation: content/en/profiler/enabling/java.md

{{< tabs >}}
{{% tab "JFR" %}}

> The JFR based allocation profiling engine is enabled by default since JDK 16. For older JDKs, manual activation and performance considerations should be taken into account.

Author: added minor clarification

> This is not supported for Oracle JDK 8 and IBM J9 JDK 8. A paid license version of Oracle JDK 8 enables the usage of JFR events on which this feature relies.

Reviewer: > A paid license version of Oracle JDK 8 enables the usage of JFR events on which this feature relies.

So I actually sat down to validate this, and as I suspected, it's not correct.

Specifically, with an OpenJDK-based distribution, in this case coretto:

```
$ java -version && env DD_PROFILING_ENABLED_EVENTS=jdk.ObjectAllocationInNewTLAB,jdk.ObjectAllocationOutsideTLAB java -javaagent:dd-java-agent.jar -Ddd.profiling.enabled=true -Ddd.profiling.ddprof.enabled=false  TestApp
openjdk version "1.8.0_442"
OpenJDK Runtime Environment Corretto-8.442.06.1 (build 1.8.0_442-b06)
OpenJDK 64-Bit Server VM Corretto-8.442.06.1 (build 25.442-b06, mixed mode
```

I do get a JFR with allocation in new TLAB / outside TLAB when I enable the events.

Switching over to Oracle JDK 8 on my local machine:

```
$ java -version && env DD_PROFILING_ENABLED_EVENTS=jdk.ObjectAllocationInNewTLAB,jdk.ObjectAllocationOutsideTLAB java -XX:+UnlockCommercialFeatures -XX:+FlightRecorder -javaagent:dd-java-agent.jar -Ddd.profiling.enabled=true -Ddd.profiling.ddprof.enabled=false  TestApp
java version "1.8.0_441"
Java(TM) SE Runtime Environment (build 1.8.0_441-b07)
Java HotSpot(TM) 64-Bit Server VM (build 25.441-b07, mixed mode)
```

... I get no events related to Allocation in TBLA/outside TLAB.

Thus, to sum it up:
* You *do* need the paid license to use JFR in Oracle JDK 8
* Oracle JDK 8 does not support allocation profiling, even if you try to enable it.

(Also, should the Java library maybe emit some info about this? I didn't get any error saying "you tried to enable the event but it's not supported...").

{{< tabs >}}
{{% tab "JFR" %}}
The JFR based allocation profiling engine is enabled by default since JDK 16.

> The reason it is not enabled by default for Open JDK 8 and JDK 11, is that an allocation intensive

Reviewer: ```suggestion
The reason it is not enabled by default for OpenJDK 8 and JDK 11, is that an allocation intensive
```

Author: thank you

> The reason it is not enabled by default for OpenJDK 8 and JDK 11, is that an allocation intensive
> application can lead to high overhead and large recording sizes. 
> This is not supported for Oracle JDK 8 and IBM J9 JDK 8. A paid license version of Oracle JDK 8
> enables the usage of JFR events on which this feature relies.

Reviewer 2: Recommend reordering so that the enabling information isn't broken up 

```suggestion

This is not supported for Oracle JDK 8 and IBM J9 JDK 8. A paid license version of Oracle JDK 8 enables the usage of JFR events on which this feature relies.

It is not enabled by default for OpenJDK 8 and JDK 11 because an allocation intensive application can lead to high overhead and large recording sizes. 

```

> enables the usage of JFR events on which this feature relies.

Reviewer: > A paid license version of Oracle JDK 8
> enables the usage of JFR events on which this feature relies.

I'm actually not sure if this is the case. May be worth checking.

Author: The usage of the `UnlockCommercialFeatures` seems to be required, though identifying what versions need this seems difficult. I do not think we want to go into this level of detail.

```
-Ddd.profiling.enabled.events=jdk.ObjectAllocationInNewTLAB,jdk.ObjectAllocationOutsideTLAB
```

> This is not supported for Oracle JDK 8 and IBM J9 JDK 8. A paid license version of Oracle JDK 8 enables this capability through an upgrade in JFR events.

Reviewer: I'm not sure the "paid license" for Oracle JDK applies to this event; I believe the "paid license" refers to using any JFR-related features in Oracle JDK 8.

(And I'm not sure if the events above exist in Oracle JDK 8...)

Reviewer: Also: maybe perhaps move this to the top, with the rest of the content?
